generators: support off-chain Python SDK

luc10921 opened the pull request.

Commit:
CU-86a0wfgjd - Implement the offchain SDK generation for Python following the template

melanke commented:
Task linked: [CU-86a0wfgjd Implement the offchain SDK generation for Python following the template](https://app.clickup.com/t/86a0wfgjd)

ixje reviewed (changes requested):
Overall it seems to generate as intended.  

The biggest thing we need to change is the output directory. Right now we write the `offchain` SDK code in the same package as the `onchain` SDK. Those should be separate. The original program structure did not consider `offchain` SDKs as a feature and so by default writes everything under `./cpm_out/<language>/<sdk>` if no custom output directory is specified. We should split that into `./cpm_out/onchain/<language>/<sdk>` and `./cpm_out/offchain/<language>/<sdk>`.

This can be handled around here
https://github.com/CityOfZion/cpm/blob/0ba454d8520711a86d5b2f95d37366b0dae839ae/config.go#L137-L143


Some smaller requests
- Please update the `generate` help message for Python to be generic. Now it still says "on-chain"
- Please update the documentation in the [tools section](https://github.com/CityOfZion/cpm/blob/master/docs/config.md#tools) to include python in the support offchain language list

Commit:
CU-86a0wfgjd - CR - add a new offchain or onchain directory on the default cpm generation path

Commit:
CU-86a0wfgjd - CR - smaller requests, change some messages to be more helpful to the user

luc10921 commented:
On [this commit](https://github.com/CityOfZion/cpm/pull/13/commits/bd729fff983cf9334f8b58eb1bcb3d8d9a466952) I also fixed a `panic: runtime error: invalid memory address or nil pointer dereference` error, it was happening because [cfg.Defaults.OnChain](https://github.com/CityOfZion/cpm/blob/0ba454d8520711a86d5b2f95d37366b0dae839ae/main.go#L450) or [cfg.Defaults.OffChain](https://github.com/CityOfZion/cpm/blob/0ba454d8520711a86d5b2f95d37366b0dae839ae/main.go#L461) could be `nil`, so I added a verification to check if it's not `nil`.

On [this other commit](https://github.com/CityOfZion/cpm/pull/13/commits/20aef48b3394557a4b913521dfc2cc091f0a2867) I changed the name of the file generate on the offchain SDK (from `contract_off_chain_sdk.py` to `contract.py`), since it will be inside the `offchain` folder, and I also added the SDK Type (`on-chain` or `off-chain`) on the error messages

ixje commented on config.go:
```diff
@@ -121,43 +121,38 @@ func (c *CPMConfig) getHosts(networkLabel string) []string {
-	// TODO: this is a bit of a hack
-	// Resolve this properly once we have a second language that can do off-chain generation or a language that supports both
-	if forLanguage == LANG_TYPESCRIPT {
-		if c.Defaults.OffChain == nil {
```
the removal of this check now causes a null pointer exception in the following scenario
* remove the "off-chain" key from the `defaults` sections in `cpm.yaml`
* run `cpm generate ts -m somecontract.manifest.json`
```go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa6d4f6]

goroutine 1 [running]:
main.(*CPMConfig).getSdkDestination(0xc00024ae10, {0xb40194, 0x2}, {0x0, 0x0})
        /home/erik/code/cpm/config.go:129 +0xf6
main.handleCliGenerate(0xc0001f2c40, {0xb40194, 0x2})
        /home/erik/code/cpm/main.go:416 +0x2e5
```

ixje commented on main.go:
```diff
@@ -435,25 +435,29 @@ func fetchManifestAndGenerateSDK(c *ContractConfig, host string) error {
-		err = generateSDK(&generators.GenerateCfg{Manifest: m, ContractHash: c.ScriptHash, SdkDestination: cfg.getSdkDestination(l)}, l, SDK_ONCHAIN)
-		if err != nil {
-			return err
+	if cfg.Defaults.OnChain != nil {
```
The config defaults are optional. This checks prevents using a `cpm.yaml` file where `on-chain` is missing from the defaults section, but where the `contract` section does specify an `on-chain` section. e.g.
```
defaults:
  contract-source-network: mainnet
  contract-destination: neo-express
  contract-generate-sdk: true
contracts:
  - label: otter
    script-hash: '0x963e74226c0cffa8a2369116f298579f9721b09d'
    on-chain:
      languages:
        - python
...
```

Without this check it would have failed on the next line (which was probably the bug you encountered). I think we need to do something like the following
```go
	var languages []string
	if c.OffChain != nil {
		languages = c.OffChain.Languages
	} else if cfg.Defaults.OffChain != nil {
		languages = cfg.Defaults.OffChain.Languages
	} else {
		return fmt.Errorf("no off-chain SDK language specified. Add the 'off-chain' key to either the 'defaults'" +
			" or 'contract' specific section of your 'cpm.yaml' file")
	}
	for _, l := range languages {
		err = generateSDK(...)
```
because the `contract` specific configuration should have priority over the `defaults`.

Commit:
CR - Change `handleCliGenerate` to not load a cpm config file and change `fetchManifestAndGenerateSDK` to check "contract" before "default" from cpm config file

Commit:
Moved SDK_ONCHAIN and SDK_OFFCHAIN from cpm/main.go to cpm/generators/common.go

Commit:
Add "off-chain" into cpm.yaml.default

ixje committed:
typo

ixje committed:
error if no SDK is generated

ixje committed:
Merge branch 'master' into CU-86a0wfgjd

ixje commented on main.go:
```diff
@@ -441,33 +446,50 @@ func handleCliVersion(cCtx *cli.Context) error {
 		}
 	}
+
+	if onChainLanguages == nil && offChainLanguages == nil {
+		return fmt.Errorf("nothing to generate. Ensure your 'cpm.yaml' has an 'onchain' or 'offchain' key under " +
+			"the 'defaults' section or contract specific section with at least one language specified")
+	}
```
I've added this check to ensure that a `cpm run` where both the `default` section and contract specific section in `cpm.yaml`  are empty generates an error because returning `nil` indicates "generation success".

ixje commented:
Thanks for the effort :1st_place_medal:
